synapse/storage/engines/postgres.py

```diff
@@ -197,17 +205,21 @@ def server_version(self):
         else:
             return "%i.%i.%i" % (numver / 10000, (numver % 10000) / 100, numver % 100)
 
-    def in_transaction(self, conn: Connection) -> bool:
-        return conn.status != self.module.extensions.STATUS_READY  # type: ignore
+    def in_transaction(self, conn: "psycopg2.connection") -> bool:
+        import psycopg2.extensions
+
+        return conn.status != psycopg2.extensions.STATUS_READY
```

Comment:
Note that we'll be running the `import` on every database transaction. I think it's fine?
Reimporting a previously imported module didn't seem to take too long in the repl:
```
>>> t0 = time.time(); import psycopg2.extensions; time.time() - t0
1.0013580322265625e-05 (worst number)
>>> t0 = time.time(); time.time() - t0
2.384185791015625e-06 (worst number)
```

Reply:
I'd call it "ugly, but fine".

The obstacle blocking us from importing `psycopg2` freely in this module is that we except to be able to import `PostgresEngine` everywhere, even if psycopg2 is not available. For example:

https://github.com/matrix-org/synapse/blob/01e625513afbd9645ff642810d559f96fab13278/synapse/storage/databases/main/events_worker.py#L181

https://github.com/matrix-org/synapse/blob/3e0536cd2afb5a640619bd872fc27b068ec3eb9b/synapse/storage/databases/state/bg_updates.py#L102

https://github.com/matrix-org/synapse/blob/75dff3dc980974960f55fa21fc8e672201f63045/synapse/storage/databases/main/relations.py#L468

I mulled over doing something like

- rename `postgres.py` to `_postgres.py`
- create a new file `postgres.py` with contents like this:
  ```python
  try:
      from ._postgres import PostgresEngine
  except ImportError:
      class PostgresEngine: pass

  __all__ = ["PostgresEngine"] 
  ```

But I'm not sure how well Pycharm and mypy will handle that. Other ideas:

- have DatabaseEngine have some kind of "engine_kind" attribute which points to an enum. Felt like duplicating information we already know, and just generally gave me a smelly impression.
- make psycopg2 a mandatory requirement for Synapse, That'd be my preferred option, but I suspect it'd be controversial.

Reply:
@DMRobertson another idea would be to add more flags to database engine that talk about the *features* as opposed to the specific class. E.g. it looks like those need (respectively):

* Sequence support
* `RECURSIVE` support
* `DISTINCT ON` support

E.g. expand the features listed near https://github.com/matrix-org/synapse/blob/86a515ccbf359ecd65a42a3f409b8f97c8f22284/synapse/storage/engines/postgres.py#L151-L166

Reply:
That's a good idea, though I think it'd be tricky to retrofit those in across the source tree without being familiar with the differences between the two databases.

(Obligatory grumble: I don't want us to be rewriting and maintaining our own version of SQLAlchemy core)

Reply:
Completely fair, not trying to scope creep! 👍 I've thought about doing this in the past also!